agent/exec/containerd/controller.go

```diff
@@ -195,12 +195,18 @@ func (r *controller) Wait(ctx context.Context) error {
 		return errors.Wrap(err, "inspecting container failed")
 	}
 
+	// TODO(ijc) this shouldn't be needed here, figure out why
+	// .shutdown/.remove are not being called otherwise.
 	shutdownWithExitStatus := func(reason string) error {
 		exitStatus, err := r.adapter.shutdown(ctx)
 		if err != nil {
 			return err
 		}
 		log.G(ctx).Errorf("EXIT STATUS %v", exitStatus)
+		if err := r.adapter.remove(ctx); err != nil {
```

Comment:
It's not correct to remove the container automatically on shutdown. We only remove it when the task is removed from the assignment set. This way old containers are kept around as long as the corresponding historic task is retained.

`Remove` should be called by `reconcileTaskState` once the task is no longer in the assignment set. With default settings, you may have to cycle through 5 tasks for a particular slot (or scale down) to see a removal happen.